src/main/scala/edu/ie3/osmogrid/lv/region_coordinator/OsmoGridModelPartitioner.scala

```diff
@@ -44,42 +44,33 @@ object OsmoGridModelPartitioner extends LazyLogging {
       osmoGridModel: LvOsmoGridModel,
       areas: ParMap[AreaKey, Polygon]
   ): ParMap[AreaKey, LvOsmoGridModel] = {
-    val buildings = assign(
-      osmoGridModel.buildings,
-      areas,
-      AssignByMax
-    )
-    val highways = assign(
-      osmoGridModel.highways,
-      areas,
-      AssignToAll
-    )
-    val landuses = assign(
-      osmoGridModel.landuses,
-      areas,
-      AssignToAll
-    )
-    val boundaries = assign(
-      osmoGridModel.boundaries,
-      areas,
-      AssignByMax
-    )
-    val existingSubstations = assign(
-      osmoGridModel.existingSubstations,
-      areas,
-      AssignByMax
-    )
-
-    areas.keys.map { areaId =>
-      areaId ->
-        LvOsmoGridModel(
-          buildings.getOrElse(areaId, ParSeq.empty),
-          highways.getOrElse(areaId, ParSeq.empty),
-          landuses.getOrElse(areaId, ParSeq.empty),
-          boundaries.getOrElse(areaId, ParSeq.empty),
-          existingSubstations.getOrElse(areaId, ParSeq.empty),
-          osmoGridModel.filter
+    val buildings = assign(osmoGridModel.buildings, areas, AssignByMax)
+    val existingSubstations =
+      assign(osmoGridModel.existingSubstations, areas, AssignByMax)
+
+    areas.keys.flatMap { areaId =>
+      val assignedBuildings = buildings.getOrElse(areaId, ParSeq.empty)
+      val assignedSubstations =
+        existingSubstations.getOrElse(areaId, ParSeq.empty)
+
+      if (assignedBuildings.nonEmpty || assignedSubstations.nonEmpty) {
+        Some(
+          areaId ->
+            LvOsmoGridModel(
+              assignedBuildings,
+              assign(osmoGridModel.highways, areas, AssignToAll)
+                .getOrElse(areaId, ParSeq.empty),
+              assign(osmoGridModel.landuses, areas, AssignToAll)
+                .getOrElse(areaId, ParSeq.empty),
+              assign(osmoGridModel.boundaries, areas, AssignByMax)
+                .getOrElse(areaId, ParSeq.empty),
```

Comment:
With this change you are building the same maps over and over. I think it would be possible to keep the assigning outside of the `flatMap`.